### ISSUE TITLE
Pythonista3

### ISSUE THREAD

**Issue by syslot** [ywangd/stash#179]
Pythonista was released，would you prepare to adapt to it with python3?

**Comment by ywangd** [comment]
You can still use StaSh with the new release. The `launch_stash.py` script contains a `python2` shebang that will allow it running in Python 2 mode even when you have Python 3 as default interpreter.
 
With above being said, I do plan to make stash python 3 compatible. In fact, the [`py3`](https://github.com/ywangd/stash/tree/py3) branch exists for this very purpose. However the work is not completed as there are some issues porting the various command scripts, especially `git` as one of its dependency is Python 2 only. 

Since new Pythonista can run both 2 & 3 code, I am thinking to port only the StaSh main programs to Python 3 and leaves the command scripts to Python 2. However, one critical missing piece is to invoke the corresponding Python interpreter from with StaSh based on the shebang line of a script. This is going to involve some objc hacks and I haven't had time to work on it. @jsbain could you please shed some light on this?

**Comment by jsbain** [comment]
I think i was able to execute a script that was already loaded in the editor, but not a generic script.   I forget.  I think I concluded it was going to be difficult to do it seamlessly from within a script.  But there may also be a way from a cdll, I have not played with that, not sure if both are accessible... 
>

**Comment by syslot** [comment]
Well ,  very thanks to your tool, and stash is real nice project!

**Comment by ywangd** [comment]
@jsbain 
I tried with `PA3PythonInterpreter` object and it can run an arbitrary script using the `runScriptWithFileAtPath_argv_` method. I am not sure how it would work when putting in a separate thread though and that's something to test. 
However, I was NOT able to get the `PythonInterpreter` object to work. There is this `runFile_argv_captureExceptions_completion_` method, but I don't know what do you pass as the third and fourth arguments. 

Since I plan to make StaSh main programs PY3 compatible, it is more important to get `PythonInterpreter` to work as most of the command scripts are PY2.

**Comment by syslot** [comment]
I try ssh & gh in pythonista3 and two scripts are broken.

**Comment by jsbain** [comment]
I Have gotten this to work through the editor tab:
https://gist.github.com/e64581caa4588625896594882870c50c
The file needn't be open, but some file needs to be open (tab must exist).  If interpreter is already running, you must wait for running to end or your request is ignored.  See code to do so (not thread safe, but a Lock could work).  
The main drawback-- globals are cleared.

I tried swizzling a few of the PythonInterpreter methods to log when they are called, so we can see what the argument types are, but the run_asFiles do not seem to get called.  Need to try again with a few others, or maybe swizzle the call selector stuff.

BTW, if you have not yet installed @dgelessus's faulthandler in his pythonista_startup repo... do so immediately!  When things crash in objc it shows the exception, etc.

**Comment by jsbain** [comment]
Aha!
``` 
I2=ObjCClass('PythonInterpreter').sharedInterpreter()
I2.runWithOptions_(({ 'captureExceptions':0,
								'script':'print "hello"\n',
								'scriptPath':"<string>"}))
```

**Comment by jsbain** [comment]
There is no obvious way to get exceptions or return codes back to py3.

**Comment by bennr01** [comment]
@jsbain Would it be possible to give each command an ID, then run the script with something like

`try:`
`    exec script in <namespace to be passed> #run script`
`except Exception as e: #catch exception`
`    dump_exception(e, os.path.join(<some_temp_dir>, command_id)) #save exception to file`

and let StaSh read the file to print the exception?

**Comment by jsbain** [comment]
Another option would be to use a socket, such as an AF_UNIX socket which talks through a file id.  

Since many scripts use _stash, on the py2 side, stash must be enabled, and when scripts are called something will need to be done to synchronize the state.  If a py2 script runs a py3 command, things might start getting dicey..

**Comment by jsbain** [comment]
actually, i guess it is possible to communicate on the objc side. Also, it feels like it should be possible using ctypes to call the c level commands for running a line, but i have not investigated yet

**Comment by ywangd** [comment]
@jsbain Thanks for all the diggings. 

The `runScriptAtPath_withInterpreterVersion_arguments_` method does NOT seem to be useful if it has to run with `on_main_thread`. This is because all stash commands runs in separate thread and is not compatible with this requirement.

**Comment by ywangd** [comment]
@syslot ssh works fine for me. What's the exact error message you had? Also maybe worthwhile to create a separate issue if you are still experiencing problems.

**Comment by jsbain** [comment]
I don't think it has to be run on the main thread. It seemed like a good idea, but might not be needed.

**Comment by dgelessus** [comment]
The standard Python C API might be useful for this too. I did some simple tests and it seemed to work properly. Maybe the C API has better ways of detecting exceptions and such than Pythonista's methods.

**Comment by jsbain** [comment]
I was thinking that too.... How do you get access to the different interpreters from that standard Python C API?  are there different static libraries to reference,  or something, or are names mangled differently?

**Comment by dgelessus** [comment]
Think I posted this on the forum a while back... you can create a `ctypes.CDLL` with the path of the `PythonistaKit3` or `PythonistaKit` executable, then you have access to the Python 3 and 2 API respectively.

**Comment by ywangd** [comment]
C API is probably more stable for the x-interpreter calls if it is usable. However, I tried with following simple code and it crashes Pythonista.
```python
import ctypes
import os
import sys

EXE_DIR = os.path.dirname(sys.executable)
p3f = 'PythonistaKit3'

dll = ctypes.CDLL(os.path.join(EXE_DIR, 'Frameworks', p3f + '.framework', p3f))

dll.PyRun_SimpleString.argtypes = [ctypes.c_char_p]
dll.PyRun_SimpleString.restype = ctypes.c_int
dll.PyRun_SimpleString('print(42)')
```

My default interpreter is 2.7. I tried similar code with Python on my laptop and it crashes as well. I have no idea where to look at. Help is highly appreciated! @dgelessus @jsbain

**Comment by dgelessus** [comment]
@ywangd This is a piece of code I wrote a while ago. It doesn't run Python code (I have code that does that too, but it's buried somewhere in the Pythonista 3 thread) but maybe you can work based on this.

```python
import ctypes
import os
import sys

FRAMEWORKS = os.path.join(os.path.dirname(sys.executable), "Frameworks")

def get_python_version(name):
	dll = ctypes.CDLL(os.path.join(FRAMEWORKS, name + ".framework", name))
	dll.Py_GetVersion.argtypes = []
	dll.Py_GetVersion.restype = ctypes.c_char_p
	return dll.Py_GetVersion().decode("ascii")

print(get_python_version("PythonistaKit"))
print(get_python_version("PythonistaKit3"))
```

**Comment by ywangd** [comment]
@dgelessus I did make my crashing code following the example you pasted. Your example did run. However it just won't allow the execution of another python script.

**Comment by dgelessus** [comment]
@ywangd You probably need to work with the other interpreter's GIL to be able to call most API functions. See [this section of the Python C API manual](https://docs.python.org/3/c-api/init.html#thread-state-and-the-global-interpreter-lock), specifically the `PyGILState_Ensure` and `PyGILState_Release` functions.

Are you in the Pythonista Slack channel? If you're not - I recently found a way to create a fully separate Python interpreter in Pythonista (by copying the `PythonistaKit3.framework`) but I haven't done much with it yet, so I don't know how practical this is.

**Comment by Dutcho** [comment]
Any progress on this?
The current version (master branch) doesn't allow python3 code, even with shebang.

**Comment by bennr01** [comment]
@Dutcho Its about 30% done. There is a experimental `python3`-command in bennr01:restructure (use `selfupdate -f bennr01:restructure` to switch to the branch), but it is still mostly bugged and untested.

**Working so far:**
- executing python3 code from stash (`python3 -c "print 2+2"` will show a SyntaxError)
- I/O (but not perfectly)
- `sys.argv` and cwd (but not really tested)
- Errors

** Known Bugs:**
- The REP-loop in the `python3`-command does not work (=`python3` without args wont work)
- Some modules executed with `python3 -m module_name_here` give weird errors (`python3 -m cmd` gives a NameError for "PROMPT")
- You need to press `return` after the `python3` command finishes in order to exit the script

**Still needs to be implemented:**:
- the `_stash` global for python3 scripts (it will probably use some sort of a RPC system)
- function to kill the python3 script
- correct exit codes
- automatic version detection using a shebang
